### PR TITLE
Issue/991 add script support to prepare

### LINKED ISSUE #991: Feature Request - Add pre and post script to package dependency install for "Prepare" command
**Request** Add the option of a pre and post script to package dependency install for "Prepare" command

**Problem statement:** We have scratch orgs failing to create with "ContactstoMultipleAccounts" feature enabled, even though we have applied the setting in the scratch org definition file. Salesforce have noted it's a Known Issue: https://trailblazer.salesforce.com/issues_view?id=a1p3A000001PYYlQAO

Given this problem and the rate our Scratch Org pools are failing at, we are looking at options to run a script prior (pre) to package dependency install in an attempt to resolve this issue.

Can this feature please be added?

### PR DESCRIPTION
**Pull request by zhebinliu**
Fixes #991 

#### Checklist
All items have to be completed before a PR is merged

- [x] Adhere to [Contribution Guidelines](https://docs.dxatscale.io/about-us/contributing-to-dx-scale)
- [x] Updates to Decision Records considered?
- [x] Updates to documentation at [DX@Scale Guide](https://github.com/dxatscale/dxatscale-guide) considered?
- [x] Tested changes?
- [x] Unit Tests new and existing passing locally?

### CODE REVIEW

File: packages/core/src/package/scriptExecutors/scriptExecutorHelpers.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import ExecuteCommand from '../../command/commandExecutor/ExecuteCommand';
+import SFPLogger, { Logger, LoggerLevel } from '../../logger/SFPLogger';
+import defaultShell from '../../utils/DefaultShell';
+
+export default class scriptExecutorHelpers {
```
Comment (azlam-abdulsalam) [inline packages/core/src/package/scriptExecutors/scriptExecutorHelpers.ts]: Can you rename this to start with Caps.. to align with other Files

Comment (azlam-abdulsalam) [reply packages/core/src/package/scriptExecutors/scriptExecutorHelpers.ts]: Also lets probably move this from package section, as this is now used elsewhere

File: packages/core/src/scratchorg/pool/PoolConfig.ts [hunk 2]
```diff
@@ -22,6 +22,8 @@ export interface PoolConfig {
         };
     };
     snapshotPool?:string;
+    postScriptPath: string;
```
Comment (azlam-abdulsalam) [inline packages/core/src/scratchorg/pool/PoolConfig.ts]: May be we make this explict, preDependencyInstallationScriptPath?

File: packages/sfpowerscripts-cli/src/impl/prepare/PrepareOrgJob.ts [hunk 3]
```diff
@@ -123,6 +129,8 @@ export default class PrepareOrgJob extends PoolJobExecutor {
                 }
             }
 
+            await this.postInstallScirpt(scratchOrg, hubOrg);
```
Comment (azlam-abdulsalam) [inline packages/sfpowerscripts-cli/src/impl/prepare/PrepareOrgJob.ts]: typo

**Review by azlam-abdulsalam — CHANGES_REQUESTED** [review]
Thanks, please have a look at the comments, Also lets call out this executes script from the head only

File: packages/core/src/scriptExecutor/ScriptExecutorHelpers.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import ExecuteCommand from '../command/commandExecutor/ExecuteCommand';
+import SFPLogger, { Logger, LoggerLevel } from '../logger/SFPLogger';
+import defaultShell from '../utils/DefaultShell';
+
+export default class scriptExecutorHelpers {
+    static async executeScript(script: string, sfdx_package: string, targetOrg: string, devHub: string, logger: Logger) {
```
Comment (azlam-abdulsalam) [inline packages/core/src/scriptExecutor/ScriptExecutorHelpers.ts]: This need to be a var args.. 
...args: string[]

File: decision records/prepare/003-prepare-custom-script-executer.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+      1. Account Contact Relation
+      2. Activity setting
+  
+
+## Decision
+
+Provide capability user can execute custom scripts before package dependency installation and after package deployment. In order to give user more control of the process, below argurements will be passed into script for use in the custom logic(same order as listed).
+
+    1. package name (not applicable in prepare commnad)
```
Comment (azlam-abdulsalam) [inline decision records/prepare/003-prepare-custom-script-executer.md]: Lets remove this from DR, as we are not sending this info

File: decision records/prepare/003-prepare-custom-script-executer.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    2. target org username
+    3. devhub username
+    4. installation/deployment status
```
Comment (azlam-abdulsalam) [inline decision records/prepare/003-prepare-custom-script-executer.md]: The code doesn't implement the 4th argument

File: decision records/prepare/003-prepare-custom-script-executer.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+### User should consider the behavior of custom script in following scenarios:
+#### The behaviour on succeedOnErrors
+Based on the 4th script arguement listed above, user should decide the behaviour on whether the script needs to be executed or have different set of code to be executed depends on the deployment status.
+
+#### The behaviour on pool definitions without request for deployment
+The scirpt should be able to read the "installAll" attribute from the config file, and build the logic about whether the script always neeed to be executed or it depends on the deployment configuration.
```
Comment (azlam-abdulsalam) [inline decision records/prepare/003-prepare-custom-script-executer.md]:
```suggestion
The script should be able to read the "installAll" attribute from the config file, and build the logic about whether the script always neeed to be executed or it depends on the deployment configuration.
```

File: decision records/prepare/003-prepare-custom-script-executer.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
+# Add custom script support in pool prepare command before dependency installation and after package installation
+
+• Status: Proposed
```
Comment (azlam-abdulsalam) [inline decision records/prepare/003-prepare-custom-script-executer.md]:
```suggestion
• Status: Accepted
```

**Review by azlam-abdulsalam — CHANGES_REQUESTED** [review]
Few more changes and we are good

File: decision records/prepare/003-prepare-custom-script-executer.md (hunk 6 above)
Comment (zhebinliu) [reply decision records/prepare/003-prepare-custom-script-executer.md]: added status to the executor

File: decision records/prepare/003-prepare-custom-script-executer.md (hunk 5 above)
Comment (zhebinliu) [reply decision records/prepare/003-prepare-custom-script-executer.md]: removed

File: packages/sfpowerscripts-cli/src/impl/prepare/PrepareOrgJob.ts [hunk 9]
```diff
@@ -273,18 +274,29 @@ export default class PrepareOrgJob extends PoolJobExecutor {
...
 
         if (fs.existsSync(this.pool.preDependencyInstallationScriptPath)) {
-            SFPLogger.log(`Executing pre script for `+ scratchOrg.alias +', script path:'+ this.pool.preDependencyInstallationScriptPath,LoggerLevel.INFO,packageLogger);
-            let args = [this.pool.postDeploymentScriptPath,null,scratchOrg.username,hubOrg.getUsername(),null];
-            await ScriptExecutor.executeScript(args, packageLogger);
+            SFPLogger.log(`Executing pre script for `+ scratchOrg.alias +', script path:'+ this.pool.preDependencyInstallationScriptPath, LoggerLevel.INFO, packageLogger);
+            await ScriptExecutor.executeScript(
+                packageLogger,
+                this.pool.preDependencyInstallationScriptPath,
+                null,
```
Comment (azlam-abdulsalam) [inline packages/sfpowerscripts-cli/src/impl/prepare/PrepareOrgJob.ts]: Can we remove null gere

**Pull request merged**